pkg/component/controller/cplb_unix.go

```diff
@@ -290,39 +317,100 @@ func (k *Keepalived) generateKeepalivedTemplate() error {
 	return nil
 }
 
+func (k *Keepalived) watchReconcilerUpdates() {
+	// Wait for the supervisor to start keepalived before
+	// watching for endpoint changes
+	process := k.supervisor.GetProcess()
+	for i := 0; process == nil; i++ {
+		if i > 5 {
+			k.log.Error("failed to start keepalived, supervisor process is nil")
+			return
+		}
+		k.log.Info("Waiting for keepalived to start")
+		time.Sleep(5 * time.Second)
+		process = k.supervisor.GetProcess()
+	}
+
+	k.log.Info("started watching cplb-reconciler updates")
+	for range k.updateCh {
+		k.keepalivedConfig.RealServers = k.reconciler.GetIPs()
+		k.log.Infof("cplb-reconciler update, got %s", k.keepalivedConfig.RealServers)
+		if err := k.generateKeepalivedTemplate(); err != nil {
+			k.log.Errorf("failed to generate keepalived template: %v", err)
+			continue
+		}
+
+		process := k.supervisor.GetProcess()
+		if err := process.Signal(syscall.SIGHUP); err != nil {
+			k.log.Errorf("failed to send SIGHUP to keepalived: %v", err)
+		}
+	}
+	k.log.Info("stopped watching cplb-reconciler updates")
+}
+
 // keepalivedConfig contains all the information required by the
 // KeepalivedConfigTemplate.
 // Right now this struct doesn't make sense right now but we need this for the
 // future virtual_server support.
 type keepalivedConfig struct {
-	VRRPInstances []k0sAPI.VRRPInstance
+	VRRPInstances  []k0sAPI.VRRPInstance
+	VirtualServers []k0sAPI.VirtualServer
+	RealServers    []string
+	APIServerPort  int
 }
 
 const dummyLinkName = "dummyvip0"
-const keepalivedConfigTemplate = `
-{{ range .VRRPInstances }}
-vrrp_instance {{ .Name }} {
+
+var keepalivedConfigTemplate = template.Must(template.New("keepalived").Parse(`
+{{ range $i, $instance := .VRRPInstances }}
+vrrp_instance k0s-vip-{{$i}} {
 	# All servers must have state BACKUP so that when a new server comes up
 	# it doesn't perform a failover. This must be combined with the priority.
     state BACKUP
     # Make sure the interface is aligned with your server's network interface
     interface {{ .Interface }}
     # The virtual router ID must be unique to each VRRP instance that you define
-    virtual_router_id {{ .VirtualRouterID }}
+    virtual_router_id {{ $instance.VirtualRouterID }}
     # All servers have the same priority so that when a new one comes up we don't
     # do a failover
     priority 200
 #   advertisement interval, 1 second by default
-    advert_int {{ .AdvertInterval }}
+    advert_int {{ $instance.AdvertInterval }}
     authentication {
         auth_type PASS
-        auth_pass {{ .AuthPass }}
+        auth_pass {{ $instance.AuthPass }}
     }
     virtual_ipaddress {
-	    {{ range .VirtualIPs }}
+	    {{ range $instance.VirtualIPs }}
 		{{ . }}
 		{{ end }}
     }
 }
 {{ end }}
-`
+
+{{ $APIServerPort := .APIServerPort }}
+{{ $RealServers := .RealServers }}
+{{ if gt (len $RealServers) 0 }}
+{{ range .VirtualServers }}
+virtual_server {{ .IPAddress }} {{ $APIServerPort }} {
+    delay_loop {{ .DelayLoop.Seconds }}
```

Comment:
```suggestion
    {{- if gt .DelayLoop.Seconds 0.0 }}
    delay_loop {{ .DelayLoop.Seconds }}
    {{- end }}
```

Reply:
We can't in keepalived it defaults to 60. I set it to 0 because I think it doesn't make sense to delay it at all in CPLB because it's added after kubernetes.default.svc is reconciled and hence has passed all the relevant health local health checks.

Reply:
> We can't in keepalived it defaults to 60.

I see. Apparently I found yet another delay_loop in the keepalived codebase :sweat_smile:

> I set it to 0 because I think it doesn't make sense to delay it at all in CPLB because it's added after kubernetes.default.svc is reconciled and hence has passed all the relevant health local health checks.

I don't fully understand the implications, but a delay_loop of 0 is not a thing in keepalived (it will use the default of 60, then):

```console
# /var/lib/k0s/bin/keepalived --dont-fork --use-file /run/k0s/keepalived.conf --no-syslog --log-console --log-detail --dump-conf -t
(/run/k0s/keepalived.conf: Line 33) number '0' outside range [0.000001, 18446744073709.551614]
(/run/k0s/keepalived.conf: Line 33) virtual server delay loop '0' invalid - ignoring
```

But letz address this in a subsequent PR :smile: